FWCore/Framework/src/StreamSchedule.cc

```diff
@@ -554,92 +556,146 @@ namespace edm {
+    Traits::preScheduleSignal(actReg_.get(), &streamContext_);
     
-    SendTerminationSignalIfException terminationSentry(actReg_.get(), &streamContext_);
     // This call takes care of the unscheduled processing.
-    workerManager_.processOneOccurrence<Traits>(ep, es, streamID_, &streamContext_, &streamContext_, cleaningUpAfterException);
+    workerManager_.setupOnDemandSystem(ep,es);
     
     ++total_events_;
-    try {
-      convertException::wrap([&]() {
-        
-        try {
-          auto waitTask = edm::make_empty_waiting_task();
-          //set count to 2 since wait_for_all requires value to not go to 0
-          waitTask->set_ref_count(2);
-          //Begin process in reverse order since TBB is last in first out for tasks.
-          for(auto it = trig_paths_.rbegin(), itEnd = trig_paths_.rend();
-              it != itEnd; ++ it) {
-            it->processOneOccurrenceAsync(waitTask.get(),ep, es, streamID_, &streamContext_);
-          }
-          waitTask->decrement_ref_count();
-          waitTask->wait_for_all();
-          
-          if(waitTask->exceptionPtr() != nullptr) {
-            std::rethrow_exception(*(waitTask->exceptionPtr()));
-          }
-          if(results_->accept()) {
-            ++total_passed_;
-          }
-
-        }
-        catch(cms::Exception& e) {
-          exception_actions::ActionCodes action = actionTable().find(e.category());
-          assert (action != exception_actions::IgnoreCompletely);
-          assert (action != exception_actions::FailPath);
-          if (action == exception_actions::SkipEvent) {
-            edm::printCmsExceptionWarning("SkipEvent", e);
-          } else {
-            throw;
-          }
-        }
-        
-        try {
-          ParentContext parentContext(&streamContext_);
-          if (results_inserter_.get()) results_inserter_->doWork<Traits>(ep, es, streamID_, parentContext, &streamContext_);
-        }
-        catch (cms::Exception & ex) {
-          ex.addContext("Calling produce method for module TriggerResultInserter");
-          std::ostringstream ost;
-          ost << "Processing " << ep.id();
-          ex.addContext(ost.str());
-          throw;
+    auto serviceToken = ServiceRegistry::instance().presentToken();
+    auto pathsDone = make_waiting_task(tbb::task::allocate_root(),
+                                          [iTask,&ep, &es, this,serviceToken](std::exception_ptr const* iPtr) mutable
+                                          {
+                                            ServiceRegistry::Operate operate(serviceToken);
+
+                                            std::exception_ptr ptr;
+                                            if(iPtr) {
+                                              ptr = *iPtr;
+                                            }
+                                            finishedPaths(ptr, std::move(iTask), ep, es);
+                                          });
```

Comment:
Is the lambda marked `mutable` only because of the `std::move(iTask)`?  If so, is it possible to pass `iTask` as an r-value reference function argument so as to avoid the `mutable` decorator, or are you restricted by TBB's interface?